Object mapper: merge enabled flag only when explicitly set

javanna opened the pull request:
Currently, when merging an object through put mapping, the default value for enabled is applied, which is true. In case the object has been previously disabled, that is going to cause an error although the updated mapping does not attempt to change the value for the enabled flag.

What needs to happen instead is that the enabled flag is applied only when set explicitly.

Closes #67972

javanna committed:
Object mapper: merge enabled flag only when explicitly set

Currently, when merging an object through put mapping, the default value for enabled is applied, which is true. In case the object has been previously disabled, that is going to cause an error although the updated mapping does not attempt to change the value for the enabled flag.

What needs to happen instead is that the enabled flag is applied only when set explicitly.

Closes #67972

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

romseygeek commented on server/src/test/java/org/elasticsearch/index/mapper/ObjectMapperMergeTests.java:
```diff
@@ -97,6 +108,18 @@ public void testMergeEnabledForRootMapper() {
+
+        RootObjectMapper merged = (RootObjectMapper) rootObjectMapper.merge(mergeWith);
+        assertFalse(merged.isEnabled());
+    }
```
Can we assert that the runtime field is still available as well?

javanna replied:
good point, I forgot that

javanna committed:
Object mapper: merge enabled flag only when explicitly set

Currently, when merging an object through put mapping, the default value for enabled is applied, which is true. In case the object has been previously disabled, that is going to cause an error although the updated mapping does not attempt to change the value for the enabled flag.

What needs to happen instead is that the enabled flag is applied only when set explicitly.

Closes #67972

romseygeek reviewed (approved):
LGTM!
